### DISCUSSION TITLE
Generic type in an inner class

### DISCUSSION THREAD

**Discussion by Sanchoyzer** [python/typing#1607] [Q&A]
I'm trying to use a generic type in a subclass, but it's failed
```python
from sqlalchemy.orm import DeclarativeBase
from typing import Generic, TypeVar


class Base(DeclarativeBase):
    pass


T = TypeVar('T', bound=Base)


class MyClass(Generic[T]):
    my_var: T  # ok

    class Meta:
        another_var: T  # error
```
mypy==1.8.0
```
main.py:16: error: Type variable "main.T" is unbound  [valid-type]
main.py:16: note: (Hint: Use "Generic[T]" or "Protocol[T]" base class to bind "T" inside a class)
main.py:16: note: (Hint: Use "T" in function signature to bind "T" inside a function)
```
but `T` has the bound, and I've used `Generic[T]` as the hint said. How can I fix it?

**Comment by erictraut** [comment]
Mypy is correct to flag this as an error. The scoping rules for type variables (as documented in the [Python typing spec](https://typing.readthedocs.io/en/latest/spec/generics.html#scoping-rules-for-type-variables)) do not allow the use of a TypeVar bound to an outer class to be used within an inner class.

**Comment by KubaO** [comment]
OK, that's great, but what is the correct way to specify this?

**Reply by hauntsaninja** [reply]
> what is the correct way to specify this

The current code doesn't really have a meaning, just do want to emphasis that it's important to understand type variable scoping in this case, otherwise you may continue to be confused

One thing you may have wanted to do instead is to make the inner class generic, and then when using the inner class in the outer class parametrise it by one of the type args the outer class is generic over. For clarity, I would recommend using PEP 695 syntax (if on Python 3.12 or newer) and different type variable names
